controllers/update_pods.go

```diff
@@ -72,14 +72,84 @@ func (updatePods) reconcile(ctx context.Context, r *FoundationDBClusterReconcile
 	return deletePodsForUpdates(ctx, r, cluster, updates, logger, status)
 }
 
+// processGroupIsUnavailable returns true if the process group is unavailable.
+func processGroupIsUnavailable(processGroupStatus *fdbv1beta2.ProcessGroupStatus) bool {
+	// If the Process Group has Pods is pending state, we count it as unavailable.
+	if processGroupStatus.GetConditionTime(fdbv1beta2.PodPending) != nil {
+		return true
+	}
+	// If the Process Group is missing processes, we count it as unavailable.
+	if processGroupStatus.GetConditionTime(fdbv1beta2.MissingProcesses) != nil {
+		return true
+	}
+	// If the Pod is running with failed containers, we count it as unavailable.
+	if processGroupStatus.GetConditionTime(fdbv1beta2.PodFailing) != nil {
+		return true
+	}
+	return false
+}
+
+// getFaultDomainsWithUnavailablePods returns a map of fault domains with unavailable Pods. The map has the fault domain as key and the value is not used.
+func getFaultDomainsWithUnavailablePods(ctx context.Context, logger logr.Logger, reconciler *FoundationDBClusterReconciler, cluster *fdbv1beta2.FoundationDBCluster) map[fdbv1beta2.FaultDomain]fdbv1beta2.None {
+	faultDomainsWithUnavailablePods := make(map[fdbv1beta2.FaultDomain]fdbv1beta2.None)
+
+	for _, processGroup := range cluster.Status.ProcessGroups {
+		// TODO(manfontan) only consider stateful processes. Because stateless processes do not change the fault tolerance of the cluster.
```

Comment:
Ok I will add it. We were wondering if it was better to do it in a separate PR since the filtering was not part of the original functionality. But I can add the filter now if that is fine.